### ISSUE TITLE
Validate encoding of embedded license during packaging

### ISSUE THREAD

**Issue by dotMorten** [NuGet/Home#7733]
## Details about Problem
When we went to publish a NuGet package with an embedded license file using the new license entry instead of the licenseUrl, we got a validation error that the text file wasn't UTF8 encoded.
![image](https://user-images.githubusercontent.com/1378165/52072178-5a4efb80-2539-11e9-9ff5-6b2d127dc914.png)

Looking at the file, it was UTF8 encoded with BOM, which I assume was the issue.

However I don't see the UTF8 requirement documented (https://docs.microsoft.com/en-us/nuget/reference/nuspec#license) and there are no packaging error.

Please let these sort of issues get validated during packaging, so we know of these issues up front, instead of being surprised by this during publishing.

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe): nuget.exe

NuGet version (x.x.x.xxx): 4.9

dotnet.exe --version (if appropriate):

VS version (if appropriate): VS2017

OS version (i.e. win10 v1607 (14393.321)): Win10 1809

Worked before? If so, with which NuGet version: No (embedded license is a new feature)

## Detailed repro steps so we can see the same problem

1. Use an embedded license text that doesn't use UTF8 encoding (or UTF-8 BOM encoding)

2. Attempt to publish the package to nuget.org.

/CC @karann-msft

**Comment by karann-msft** [comment]
@dotMorten how was this license file created?

**Comment by dotMorten** [comment]
@karann-msft I don't know - our legal department did this. But I don't think that's relevant. If a nuget package is going to fail server validation, I'd prefer you get similar validation client-side (warning or error).

Also I don't believe the format of the license file is documented either (I couldn't find it). There's an example that it's a `.txt` file, but nothing stating it can't be a PDF, RTF, HTML, Markdown or something else. And even if it's obvious it's a text file, the validation seems to indicate the encoding is extremely specific.

**Comment by karann-msft** [comment]
@dotMorten, it is relevant because it'll help us repro the scenario. At the moment, we are not able to repro the failure you talk about. We tried to publish a package with UTF-8 BOM encoding and it didn't fail. 

The docs have been updated to call out the supported file types - https://docs.microsoft.com/en-us/nuget/reference/nuspec#license

**Comment by dotMorten** [comment]
Didn't you get my repro on email?

**Comment by dotMorten** [comment]
Here's a package the reproduces the problem: (remove the .zip extension)
[MyPackageEsriIncTest.1.0.4.nupkg.zip](https://github.com/NuGet/Home/files/2818300/MyPackageEsriIncTest.1.0.4.nupkg.zip)

**Comment by dotMorten** [comment]
@karann-msft I've tried a bunch of different ways trying to convert or re-create the txt file, and I'm met with the same UTF8 error over and over. The only way I can get it to pass the check is I remove most of the text from the file. I'm guessing there's something in that text file that's throwing off your validator.

**Comment by dotMorten** [comment]
OK I'm finding the offending character at line 295. That unblocks us - however still I'd like to get the same validation client-side during packaging.

**Comment by karann-msft** [comment]
Can you share a file that I can repro that with?

**Comment by dotMorten** [comment]
@karann-msft Did you see package I attached 4 comments up? It has the offending `EULA.txt` file inside it.

**Comment by karann-msft** [comment]
@dotMorten - looking at it now. @rrelyea can we investigate adding pack time validations here. There does seem to be a non-UTF-8 character in the file.

**Comment by agr** [comment]
I would like to point out that Gallery accepting UTF-8 plain text files is a server implementation restriction/detail. Other implementations may choose to accept licenses in, say, pdf or doc files; or they can have a different opinion on what constitutes a "UTF-8 plain text file" (Gallery has a very relaxed check).

Adding that validation to the client might make authoring packages for anything but Gallery complicated.

**Comment by loic-sharma** [comment]
> I would like to point out that Gallery accepting UTF-8 plain text files is a server implementation restriction/detail. Other implementations may choose to accept licenses in, say, pdf or doc files

I think that having a unified experience across the NuGet client and Gallery is worth limiting the flexibility of server implementations. Servers that want that flexibility can always accept packages that use the `licenseUrl` metadata.

> Adding that validation to the client might make authoring packages for anything but Gallery complicated.

What makes you say this? The NuGet client has error codes, messages, and documentation to help users around these kind of issues. If anything, I would say that the authoring experience is complicated right now: you have to upload a package with your license to verify your license is valid.

**Comment by dotMorten** [comment]
I agree. At the very least, make it a packaging warning, if you're reluctant to make it an error.
I assume the UTF8 limitation has been made for a very good reason - probably to ensure any client consuming the package feed can understand and display the license file. If that's the case, I think it's fair to warn users that your package might cause problems for people, and not just for the NuGet Gallery.
